p2p/chain.py

```diff
@@ -244,22 +253,39 @@ def register_peer(self, peer: BasePeer) -> None:
             self, peer: HeaderRequestingPeer, start_at: int) -> Tuple[BlockHeader, ...]:
         """Fetch a batch of headers starting at start_at and return the ones we're missing."""
         self.logger.debug("Fetching chain segment starting at #%d", start_at)
-        peer.request_block_headers(start_at, peer.max_headers_fetch, reverse=False)
+        request = peer.request_block_headers(
+            start_at,
+            peer.max_headers_fetch,
+            skip=0,
+            reverse=False,
+        )
+
         # Pass the peer's token to self.wait() because we want to abort if either we
         # or the peer terminates.
-        headers = list(await self.wait(
+        headers = tuple(await self.wait(
             self._new_headers.get(),
             token=peer.cancel_token,
             timeout=self._reply_timeout))
-        for header in headers.copy():
-            try:
-                await self.wait(self.db.coro_get_block_header_by_hash(header.hash))
-            except HeaderNotFound:
-                break
-            else:
-                self.logger.debug("Discarding %s as we already have it", header)
-                headers.remove(header)
-        return tuple(headers)
+
+        # check that the response headers are a valid match for our
+        # requested headers.
+        request.validate_headers(headers)
+
+        # the inner list comprehension is required to get python to evaluate
+        # the asynchronous comprehension
+        missing_headers = tuple([
+            header
+            for header
+            in headers
+            if not (await self.wait(self.db.coro_header_exists(header.hash)))
+        ])
+        if len(missing_headers) != len(headers):
+            self.logger.debug(
+                "Discarding %d / %d headers that we already have",
+                len(headers) - len(missing_headers),
+                len(headers),
+            )
+        return headers
```

Comment:
Should this return `missing_headers`?

Reply:
Yeah, looking at these now the logic isn't nearly equivalent, I'll clean it up in #1089

Reply:
Updated here with a comment to make it clearer what the logic of the loop is doing.

https://github.com/ethereum/py-evm/pull/1089/commits/3996fcd695153a533fb43801440556f10adefa1c